File: medium/cpp/1302.cpp
// https://leetcode.com/problems/deepest-leaves-sum/description/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution
{
public:
    int deepestLeavesSum(TreeNode* root)
    {
        std::queue<std::pair<TreeNode*, int>> q;
        q.push({root, 0 });

        auto  level = 0;
        auto ans = 0;
        
        while (!q.empty())
        {
            const auto v = q.front();
            q.pop();
            
            if (v.second == level)
            {
                ans += v.first->val;
            }
            else if (v.second > level)
            {
                level = v.second;
                ans = v.first->val;
            }
            
            if (v.first->left != nullptr)
            {
                q.push({ v.first->left, v.second + 1 });
            }
            
            if (v.first->right != nullptr)
            {
                q.push({ v.first->right, v.second + 1 });
            }
        }
        
        return ans;
    }
};

File: medium/python/1302.py
# https://leetcode.com/problems/deepest-leaves-sum/description/

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def deepestLeavesSum(self, root: Optional[TreeNode]) -> int:
        q = deque([root])

        while q:
            cur_q = q
            q = deque()

            for node in cur_q:
                if node.left:
                    q.append(node.left)
                if node.right:
                    q.append(node.right)

        return sum(node.val for node in cur_q)
        